starlette/concurrency.py

```diff
@@ -1,43 +1,42 @@
-import asyncio
 import functools
-import sys
 import typing
 from typing import Any, AsyncGenerator, Iterator
 
+import anyio
+
 try:
     import contextvars  # Python 3.7+ only or via contextvars backport.
 except ImportError:  # pragma: no cover
     contextvars = None  # type: ignore
 
-if sys.version_info >= (3, 7):  # pragma: no cover
-    from asyncio import create_task
-else:  # pragma: no cover
-    from asyncio import ensure_future as create_task
 
 T = typing.TypeVar("T")
 
 
 async def run_until_first_complete(*args: typing.Tuple[typing.Callable, dict]) -> None:
-    tasks = [create_task(handler(**kwargs)) for handler, kwargs in args]
-    (done, pending) = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
-    [task.cancel() for task in pending]
-    [task.result() for task in done]
+    async with anyio.create_task_group() as task_group:
+
+        async def task(_handler: typing.Callable, _kwargs: dict) -> Any:
+            await _handler(**_kwargs)
+            task_group.cancel_scope.cancel()
+
+        for handler, kwargs in args:
+            task_group.start_soon(task, handler, kwargs)
```

Comment:
I think this helps highlight a miserable bit of code-smell.
Might be worth reviewing where we're using `run_until_first_complete`, because it's icky as heck, whichever way we slice it.

Reply:
It's only used in one other place (`starlette/responses.py`) so I can replace it there.